src/coreclr/vm/jitinterface.cpp

```diff
@@ -5990,9 +5990,10 @@ CorInfoHelpFunc CEEInfo::getCastingHelperStatic(TypeHandle clsHnd, bool fThrowin
         helper = CORINFO_HELP_ISINSTANCEOFARRAY;
     }
     else
-    if (!clsHnd.IsTypeDesc() && !Nullable::IsNullableType(clsHnd))
+    if (!clsHnd.IsTypeDesc() && !(Nullable::IsNullableType(clsHnd) && fThrowing))
     {
         // If it is a non-variant class, use the fast class helper
+        // Also use fast helper for isinst Nullable, which is equal to isinst of underlying type
```

Comment:
I'm not sure where to handle this. JIT is passing the token of original nullable type. Is there any way to get the token of underlying type?

Reply:
I mean to get it in JIT instead of VM.

I didn't update jitinterface for native aot (if exists). If done in JIT, there's no need to update in two places.

Reply:
um.. `getTypeForBox` that you already using?

Reply:
It returns a VM handle, but the JIT interface requires a metadata handle, if I'm correct.
Or we can skip the call to JIT interface and hard code to exact type helper when nullable case detected. Is there any risk?

Reply:
Presumably, you need to check that you don't deal with `Nullable<MyValueType<_Canon>>`

Reply:
```csharp
public class Program
{
    public static void Main()
    {
        Test<string>(new MyStruct<string>());
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static bool Test<T>(object o) => o is MyStruct<T>?;
}

public struct MyStruct<T>
{
    public T val;
}
```

Reply:
> `CORINFO_RESOLVED_TOKEN` to underlying type

I am not aware if we have an API for that, if we don't you will have to add it

Reply:
Since the implementation of `getCastingHelper` are all converting it to vm handle immediately, I think it's better to update jit interface to accept `CORINFO_CLASS_HANDLE`.